File: Populating Next Right Pointers in Each Node.cpp
/**
 * Definition for binary tree with next pointer.
 * struct TreeLinkNode {
 *  int val;
 *  TreeLinkNode *left, *right, *next;
 *  TreeLinkNode(int x) : val(x), left(NULL), right(NULL), next(NULL) {}
 * };
 */
class Solution {
public:
    void connect(TreeLinkNode *root) {
        if (!root)
            return;
        TreeLinkNode* pre_level = root;
        while (pre_level->left) {
            TreeLinkNode* pre_node = pre_level;
            TreeLinkNode superhead(0);
            TreeLinkNode* node = &superhead;
            while (pre_node) {
                node->next = pre_node->left;
                node = node->next;
                node->next = pre_node->right;
                node = node->next;
                pre_node = pre_node->next;
            }
            pre_level = pre_level->left;
        }
    }
};

File: Populating Next Right Pointers in Each Node.java
/**
 * Definition for binary tree with next pointer.
 * public class TreeLinkNode {
 *     int val;
 *     TreeLinkNode left, right, next;
 *     TreeLinkNode(int x) { val = x; }
 * }
 */
public class Solution {
    public void connect(TreeLinkNode root) {
        TreeLinkNode prevLevel = root;
        
        if (root == null)
            return;
        TreeLinkNode currLevel = root.left;
        
        while (currLevel != null) {
            TreeLinkNode prev = null;
            TreeLinkNode node = prevLevel;
            while (node != null) {
                if (node.left != null) {
                    if (prev != null)
                        prev.next = node.left;
                    prev = node.left;
                    if (node.right != null) {
                        prev.next = node.right;
                        prev = node.right;
                    }
                }
                node = node.next;
            }
            
            prevLevel = currLevel;
            currLevel = currLevel.left;
        }
    }
}